generators/server/templates/src/main/kotlin/package/config/SecurityConfiguration.kt.ejs

```diff
@@ -242,47 +249,56 @@ class SecurityConfiguration(
             .oauth2Login()
             <%_ } _%>
         .and()
-            .oauth2ResourceServer().jwt()
+            .oauth2ResourceServer()
+                .jwt()
+                .jwtAuthenticationConverter(mapGroupOrRolesClaimToGrantedAuthorities())
         <%_ } _%>
     }
     <%_ if (authenticationType === 'jwt') { _%>
 
-    private fun securityConfigurerAdapter(): JWTConfigurer {
-        return JWTConfigurer(tokenProvider)
-    }
+    private fun securityConfigurerAdapter() = JWTConfigurer(tokenProvider)
     <%_ } _%>
     <%_ if (authenticationType === 'oauth2') { _%>
 
+    <%_ if (applicationType !== 'microservice') { _%>
     /**
      * Map authorities from "groups" or "roles" claim in ID Token.
      *
      * @return a [GrantedAuthoritiesMapper] that maps groups from
      * the IdP to Spring Security Authorities.
      */
-    @Suppress("UNCHECKED_CAST")
     @Bean
-    fun userAuthoritiesMapper(): GrantedAuthoritiesMapper {
-        return GrantedAuthoritiesMapper { authorities ->
+    fun userAuthoritiesMapper() =
+        GrantedAuthoritiesMapper { authorities ->
             val mappedAuthorities = mutableSetOf<GrantedAuthority>()
 
             authorities.forEach { authority ->
                 val oidcUserAuthority = authority as OidcUserAuthority
-                val userInfo = oidcUserAuthority.userInfo
-                var groups = userInfo.claims["groups"] as Collection<String>?
-                if (groups == null) {
-                    groups = userInfo.claims["roles"] as Collection<String>?
-                }
-                groups?.let {
-                    mappedAuthorities.addAll(
-                        it.asSequence()
-                            .filter { group -> group.startsWith("ROLE_") }
-                            .map { a -> SimpleGrantedAuthority(a) }
+                mappedAuthorities.addAll(
+                    mapRolesToGrantedAuthorities(
+                        getRolesFromClaims(oidcUserAuthority.userInfo.claims)
                     )
-                }
+                )
             }
             mappedAuthorities
         }
-    }
+    <%_ } _%>
+
+    private fun mapGroupOrRolesClaimToGrantedAuthorities() =
+        object : JwtAuthenticationConverter() {
+            override fun extractAuthorities(jwt: Jwt) =
+                mapRolesToGrantedAuthorities(getRolesFromClaims(jwt.claims))
+        }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun getRolesFromClaims(claims: Map<String, Any?>): Collection<String> =
+        claims.getOrDefault("groups", claims.getOrDefault("roles", listOf<String>())) as Collection<String>
+
+    private fun mapRolesToGrantedAuthorities(roles: Collection<String>) =
+        roles.asSequence()
+            .filter { role -> role.startsWith("ROLE_") }
```

Comment:
I like `it` here 😄

Reply:
Yes, did several conversions to single expressions while migrating but there are still several left

Reply:
could also be reduced to:
```
.filter { it.startsWith("ROLE_") }
```

Reply:
Simplified a bit more :smile: